core/trino-main/src/main/java/io/trino/sql/planner/EffectivePredicateExtractor.java

```diff
@@ -414,6 +426,14 @@ public Expression visitValues(ValuesNode node, Void context)
                             hasNull[i] = true;
                         }
                         else {
+                            if ((types.get(node.getOutputSymbols().get(i)) instanceof RowType)
+                                    || (types.get(node.getOutputSymbols().get(i)) instanceof ArrayType)) {
+                                for (int j = 0; j < ((Block) item).getPositionCount(); j++) {
+                                    if (((Block) item).isNull(j) || ((Block) item).getSingleValueBlock(j) instanceof RowBlock) {
```

Comment:
This doesn't handle the case where the element is an array that contains nulls:

```sql
SELECT *
FROM (VALUES cast(ROW(ROW(1, array[null])) AS ROW(row(a integer, b array(integer))))) t(r)
JOIN (VALUES 42) u(c) ON c = r.a;
```

Also, the check is overly conservative. If the item is a row that doesn't contain nulls, it should no cause a problem.

What this code should do is check whether the items are:
1) null, or
2) either row or array, and that they contain at least one element that matches 1) or 2) recursively